### PR TITLE
Add retrying of failed driver initialization

### PR DESCRIPTION
**Pull request by BeyondEvil**
Fixes: #219

### CODE REVIEW

File: testing/test_driver.py [hunk 1]
```diff
@@ -180,3 +181,70 @@ def test_pass(driver_kwargs):
...
+
+        class MockDriver():
+            def __init__(*args, **kwargs):
+                pass
+            def quit(*args, **kwargs):
+                pass
+
+        @pytest.fixture(scope="session")
+        def driver_class():
+            return MockDriver
```
Comment (BeyondEvil) [inline testing/test_driver.py:200]: I tried using the mocker to avoid having these, but couldn't figure out how.

I tried:

```python
mocker.patch("pytest_selenium.pytest_selenium.webdriver")
```

and `"pytest_selenium.pytest_selenium.webdriver.Firefox"`

but couldn't get it to work 🤷‍♂️

Comment (Dude-X) [reply testing/test_driver.py:200]: I don't have too much experience with using `mock` but I think a better test is to use `Remote` webdriver, and then  mock the `RemoteConnection` class to throw an exception. This would be more realistic than having Firefox fail. Either way, focus on remote connection failing.

Comment (BeyondEvil) [reply testing/test_driver.py:200]: That's genius! 

But how do I accomplish that?

I tried
```python
    mocker.patch(
        "pytest_selenium.pytest_selenium.webdriver.remote.remote_connection.RemoteConnection",
        side_effect=Exception("Connection Error"),
    )
```

But it doesn't work. With that I mean, I don't get any syntax errors - but the test doesn't produce the expected outcome (the exception isn't raised.

Comment (BeyondEvil) [reply testing/test_driver.py:200]: I'm struggling to get something to work.

I managed to mock the `execute` method on `RemoteConnection`, but that slowed down the tests (from ~1s to ~6s).

Maybe @nicoddemus can share some wisdom?

Comment (BeyondEvil) [reply testing/test_driver.py]: Think I solved it. 👍

Comment (nicoddemus) [reply testing/test_driver.py]: Sorry I didn't get here sooner.

Was the fix moving construction of `driver_class` inside the `with` statement?

Comment (BeyondEvil) [reply testing/test_driver.py]: No worries @nicoddemus Sorry I squashed the commits and hid the history.

But no, the fix was doing the patching in the proper places.

**Comment by BeyondEvil** [comment]
ping @Dude-X @nicoddemus

File: testing/test_driver.py [hunk 2]
```diff
@@ -180,3 +183,60 @@ def test_pass(driver_kwargs):
...
+        import pytest
+
+        @pytest.mark.nondestructive
+        def test_pass(driver):
+            assert True  # catch if this starts to pass
+        """
+    )
+
+    testdir.quick_qa("--driver", "Firefox", file_test, failed=1)
+    assert (
```
Comment (nicoddemus) [inline testing/test_driver.py:238]: Minor: this was a bit hard to parse because it ended up being formatted by black, perhaps it could be extracted to a local variable:

```python
expected_attempts = max_num_attempts if max_num_attempts != "default" else 3
assert mock_retrying.spy_return.statistics["attempt_number"] == expected_attempts
```

Comment (BeyondEvil) [reply testing/test_driver.py]: Good idea!

I refined it a little bit, hope you don't mind. 😊 @nicoddemus
